### PR TITLE
Issue 547: Refactor release into subtasks.

### PR DESCRIPTION
**Pull request by mattwagl**
The most important refactorings discussed in thenativeweb/org#547 have been accomplished.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -727,7 +727,13 @@ To run the tests run the following command.
 $ npm run test
 ```
 
-You can run only a single integration test by specifing an individual test case as an additional argument.
+You can all integration test for one task by specifing the task name as an addtional argument.
```
Comment (goloroden) [inline README.md]: Somehow I didn't get this sentence. I think the verb `run` is missing here, but additionally, I didn't get what you mean by `for one task`. In the word `additional` an `i` is missing.

File: README.md [hunk 2]
```diff
@@ -727,7 +727,13 @@ To run the tests run the following command.
...
+
+```bash
+$ npm run test release
+```
+
+You can run a single integration test case by specifing the individual test case as an additional argument.
```
Comment (goloroden) [inline README.md]: In the word `specifying` a `y` is missing.

Maybe we could skip the word `case` here, as it actually is a single test, not a specific test case.

File: lib/automation/documentation/generateTocForMarkdown.js [hunk 3]
```diff
@@ -2,12 +2,17 @@
...
-const generateTocForMarkdown = async function ({ path }) {
+const generateTocForMarkdown = async function ({ path, ui }) {
   if (!path) {
     throw new Error('Path is missing.');
   }
+  if (!ui) {
+    throw new Error('UI is missing.');
+  }
+
+  ui.printTaskHeader('Readme.md TOC generation');
```
Comment (goloroden) [inline lib/automation/documentation/generateTocForMarkdown.js]: `Readme.md` => `README.md`

File: lib/automation/git/tag.js [hunk 4]
```diff
@@ -1,14 +1,19 @@
...
     throw new Error('Directory is missing.');
   }
   if (!name) {
     throw new Error('Name is missing.');
   }
+  if (!ui) {
+    throw new Error('UI is missing.');
+  }
+
+  ui.printTaskHeader('tag');
```
Comment (goloroden) [inline lib/automation/git/tag.js]: Maybe this should be `git tag`, analogously to `git push`?

File: lib/automation/pkg/checkForUnusedDependencies.js [hunk 5]
```diff
@@ -28,25 +34,36 @@ const checkForUnusedDependencies = async function ({ directory }) {
...
       if (invalidFiles.length > 0) {
-        buntstift.warn(`Unable to parse some files: ${invalidFiles.join(', ')}`);
+        hasErrorOccured = true;
+        ui.info(`Unable to parse some files: ${invalidFiles.join(', ')}`);
       }
 
       if (invalidDirs.length > 0) {
-        buntstift.warn(`Unable to access some dirs: ${invalidDirs.join(', ')}`);
+        hasErrorOccured = true;
+        ui.info(`Unable to access some dirs: ${invalidDirs.join(', ')}`);
```
Comment (goloroden) [inline lib/automation/pkg/checkForUnusedDependencies.js]: `dirs` => `directories` (in the message text)

File: lib/automation/pkg/checkForUnusedDependencies.js [hunk 6]
```diff
@@ -28,25 +34,36 @@ const checkForUnusedDependencies = async function ({ directory }) {
...
       }
 
       if (dependencies.length > 0 || devDependencies.length > 0) {
+        hasErrorOccured = true;
         const allUnusedDeps = dependencies.concat(devDependencies);
 
-        buntstift.warn(`You have unused dependencies:\n\n${allUnusedDeps.join(',\n')}`);
+        ui.info(`Following dependencies are unused or missing:\n\n${allUnusedDeps.join(',\n')}`);
```
Comment (goloroden) [inline lib/automation/pkg/checkForUnusedDependencies.js]: I know that we removed the word `The` this morning, but this way the sentence sounds strange … I'd vote for `The following dependencies …`

File: lib/automation/pkg/checkForUnusedDependencies.js [hunk 7]
```diff
@@ -28,25 +34,36 @@ const checkForUnusedDependencies = async function ({ directory }) {
...
+      }
+
+      if (hasErrorOccured) {
+        ui.warn('You have unused or missing dependencies.');
```
Comment (goloroden) [inline lib/automation/pkg/checkForUnusedDependencies.js]: `You have …` or `There are …`?

File: lib/automation/pkg/checkLicenseCompatibility.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+      table.push([
+        incompatiblePackage.name,
+        incompatiblePackage.version,
+        incompatiblePackage.license,
+        incompatiblePackage.path
+      ]);
+    });
+
+    ui.table(table);
+    ui.warn('You have incompatible dependencies.');
```
Comment (goloroden) [inline lib/automation/pkg/checkLicenseCompatibility.js]: `You have …` or `There are …`?

File: lib/automation/quality/testCode.js [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    throw new Error('UI is missing.');
+  }
+
+  ui.printTaskHeader('tests');
+
+  const testDirectory = path.join(directory, 'test');
+  const doesTestDirectoryExist = await files.exists({ path: testDirectory });
+
+  if (!doesTestDirectoryExist) {
+    ui.printTaskOmition('No test directory found, skipping tests.');
```
Comment (goloroden) [inline lib/automation/quality/testCode.js]: `omition` should be `omission` (the word `omition` does not exist 😉)

File: lib/automation/quality/testCode.js [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+testCode.getFiles = async function ({ directory }) {
+  if (!directory) {
+    throw new Error('Directory is missing.');
+  }
+
+  const sourceFiles = await files.get({
+    directory,
+    pattern: [
+      '**/*.js',
```
Comment (goloroden) [inline lib/automation/quality/testCode.js]: Shouldn't we add `**/*.jsx` here as well?

File: lib/cli/commands/deps.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+    if (help) {
+      return ui.printUsage([
+        { header: 'roboter deps', content: this.description },
+        { header: 'Synopsis', content: 'roboter deps' },
+        { header: 'Options', optionList: [ ...await this.getOptionDefinitions(), ...globalOptionDefinitions ]}
+      ]);
+    }
+
+    ui.printTaskHeader('depedency check');
```
Comment (goloroden) [inline lib/cli/commands/deps.js]: `dependency`: There is an `n` missing.

File: lib/cli/commands/release.js [hunk 12]
```diff
@@ -70,78 +64,112 @@ const release = {
...
+      ui.progress.hide();
+      ui.printTaskSuccess('Release successful.');
     } catch (ex) {
-      stopWaiting();
+      ui.progress.hide();
 
       switch (ex.code) {
         case 'ELOCALMASTERBEHINDREMOTE':
-          buntstift.error('The local master branch is behind origin/master, run git pull first.');
+          ui.printTaskFailure('The local master branch is behind origin/master, run git pull first.');
```
Comment (goloroden) [inline lib/cli/commands/release.js]: We should either add quotes around `git pull` here (as a few lines below with `npm init`) – or we should remove them there. I'd vote for adding quotes, but I'd favour single quotes over double quotes here.

File: lib/cli/ui/error.js [hunk 13]
```diff
@@ -0,0 +1,13 @@
+'use strict';
+
+const buntstift = require('buntstift');
+
+const error = function (err) {
+  if (!err) {
+    throw new Error('Error is missing');
```
Comment (goloroden) [inline lib/cli/ui/error.js]: There is the trailing `.` missing in the error message.

File: lib/cli/ui/handleException.js [hunk 14]
```diff
@@ -0,0 +1,22 @@
+'use strict';
+
+const buntstift = require('buntstift');
+
+const handleException = function (ex, messages) {
+  if (!ex) {
+    throw new Error('Exception is missing.');
+  }
+  if (!messages) {
+    throw new Error('Message are missing.');
```
Comment (goloroden) [inline lib/cli/ui/handleException.js]: `Messages` is missing the trailing `s`.

File: lib/cli/ui/info.js [hunk 15]
```diff
@@ -0,0 +1,13 @@
+'use strict';
+
+const buntstift = require('buntstift');
+
+const info = function (message) {
+  if (!message) {
+    throw new Error('Info is missing');
```
Comment (goloroden) [inline lib/cli/ui/info.js]: The trailing `.` is missing in the error message.

File: lib/cli/ui/printTaskHeader.js [hunk 16]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+
+const buntstift = require('buntstift');
+
+const printTaskHeader = function (name) {
+  if (!name) {
+    throw new Error('Name is missing');
```
Comment (goloroden) [inline lib/cli/ui/printTaskHeader.js]: The trailing `.` is missing in the error message.

File: lib/cli/ui/printTaskOmition.js [hunk 17]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+
+const buntstift = require('buntstift');
+
+const printTaskOmition = function (reason) {
+  if (!reason) {
+    throw new Error('Reason is missing');
```
Comment (goloroden) [inline lib/cli/ui/printTaskOmition.js]: The trailing `.` is missing in the error message.

File: lib/cli/ui/printTaskSuccess.js [hunk 18]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+
+const buntstift = require('buntstift');
+
+const printTaskSuccess = function (message) {
+  if (!message) {
+    throw new Error('message is missing');
```
Comment (goloroden) [inline lib/cli/ui/printTaskSuccess.js]: The trailing `.` is missing in the error message.

File: lib/cli/ui/table.js [hunk 19]
```diff
@@ -0,0 +1,13 @@
+'use strict';
+
+const buntstift = require('buntstift');
+
+const table = function (data) {
+  if (!data) {
+    throw new Error('Data is missing');
```
Comment (goloroden) [inline lib/cli/ui/table.js]: The trailing `.` is missing in the error message.

File: lib/cli/ui/warn.js [hunk 20]
```diff
@@ -0,0 +1,13 @@
+'use strict';
+
+const buntstift = require('buntstift');
+
+const warn = function (warning) {
+  if (!warning) {
+    throw new Error('Warning is missing');
```
Comment (goloroden) [inline lib/cli/ui/warn.js]: The trailing `.` is missing in the error message.

File: test/integration/deps/warns-on-unused-dependency/expected.js [hunk 21]
```diff
@@ -2,10 +2,10 @@
 
 const exitCode = 0;
 
-const stdout = ``;
-
-const stderr = `▻ You have unused dependencies:
+const stdout = `Following dependencies are unused or missing:
```
Comment (goloroden) [inline test/integration/deps/warns-on-unused-dependency/expected.js]: If we change the original error message, this should also be `The following …`.

File: test/integration/qa/warns-on-unused-dependencies/expected.js [hunk 22]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+
+const exitCode = 0;
+
+const stdout = [
+  'Running unused or missing dependencies check',
+  `Following dependencies are unused or missing:
```
Comment (goloroden) [inline test/integration/qa/warns-on-unused-dependencies/expected.js]: And again, `The following…`

File: test/integration/release/warns-on-unused-dependencies/expected.js [hunk 23]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+
+const exitCode = 0;
+
+const stdout = [
+  'Running unused or missing dependencies check',
+  `Following dependencies are unused or missing:
```
Comment (goloroden) [inline test/integration/release/warns-on-unused-dependencies/expected.js]: `The following…` 😉

File: lib/automation/quality/testCode.js (hunk 10 above)
Comment (mattwagl) [reply lib/automation/quality/testCode.js]: By default mocha can not deal with `jsx` files as the need to be compiled first. If we want to support `jsx` files we need to add `babel-register` as a dependency and tell mocha to compile jsx via the `compilers` option, see [this article](https://eddyerburgh.me/test-js-and-jsx-extensions-mocha). It is something that could be accomplished in a project with the help of mocha.opts. Hmmm…

Comment (goloroden) [reply lib/automation/quality/testCode.js]: Okay, I didn't have that in mind. Let's skip it for now.

**Comment by mattwagl** [comment]
I've included your feedback. Thanks for fixing all my damn typos!

File: README.md [hunk 24]
```diff
@@ -727,13 +727,13 @@ To run the tests run the following command.
 $ npm run test
 ```
 
-You can all integration test for one task by specifing the task name as an addtional argument.
+You can run all integration test for one task by specifying the task name as an addtional argument.
```
Comment (goloroden) [inline README.md]: There is an `i` missing in `additional` 😉

**Pull request merged**